File: mvc-sample/src/main/kotlin/suzp1984/github/io/jbig_kotlin/db/RealmDbOpenHelper.java
package suzp1984.github.io.jbig_kotlin.db;

import android.util.Log;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import javax.inject.Inject;

import io.realm.Realm;
import io.realm.RealmQuery;
import io.realm.RealmResults;
import suzp1984.github.io.jbig_kotlin.realmobj.JbigItem;

/**
 * Created by suzhenxi on 11/28/2016.
 */

public class RealmDbOpenHelper implements DataBaseHelper {

    private Realm mRealm;

    @Inject
    public RealmDbOpenHelper(Realm realm) {
        mRealm = realm;
    }

    @Override
    public List<byte[]> getJbigs() {
        List<byte[]> ret = new ArrayList<>();

        RealmQuery<JbigItem> query = mRealm.where(JbigItem.class);
        RealmResults<JbigItem> results = query.findAll();

        for (JbigItem item : results) {
            ret.add(item.getJbig());
        }

        return ret;
    }

    @Override
    public byte[] getJbig(int position) {
        RealmQuery<JbigItem> query = mRealm.where(JbigItem.class);
        RealmResults<JbigItem> results = query.findAll();

        JbigItem item = results.get(position);

        if (item != null) {
            return item.getJbig();
        }

        return null;
    }

    @Override
    public void put(byte[] jbig) {
        Log.d("ReamlDb", "save realm object.");

        mRealm.beginTransaction();

        JbigItem item = mRealm.createObject(JbigItem.class);
        item.setTag("PaintView");
        item.setJbig(jbig);

        mRealm.commitTransaction();
    }

    @Override
    public void put(Collection<byte[]> jbigs) {
        for(byte[] item : jbigs) {
            put(item);
        }
    }

    @Override
    public void delete(byte[] jbig) {
        RealmQuery<JbigItem> query = mRealm.where(JbigItem.class);
        RealmResults<JbigItem> results = query.findAll();

        for (JbigItem item : results) {
            if (item.getJbig() == jbig) {
                mRealm.beginTransaction();
                item.deleteFromRealm();
                mRealm.commitTransaction();
                break;
            }
        }
    }

    @Override
    public void delete(int position) {
        RealmQuery<JbigItem> query = mRealm.where(JbigItem.class);
        RealmResults<JbigItem> results = query.findAll();

        results.remove(position);

        mRealm.commitTransaction();
    }

    @Override
    public void deleteAll() {
        RealmQuery<JbigItem> query = mRealm.where(JbigItem.class);
        RealmResults<JbigItem> results = query.findAll();

        results.clear();
        mRealm.commitTransaction();
    }
}

File: mvc-sample/src/main/kotlin/suzp1984/github/io/jbig_kotlin/db/RealmDbOpenHelper.kt
package suzp1984.github.io.jbig_kotlin.db

import android.util.Log

import java.util.ArrayList

import javax.inject.Inject

import io.realm.Realm
import io.realm.RealmQuery
import io.realm.RealmResults
import suzp1984.github.io.jbig_kotlin.realmobj.JbigItem

/**
 * Created by suzhenxi on 11/28/2016.
 */

class RealmDbOpenHelper
@Inject
constructor(private val mRealm: Realm) : DataBaseHelper {

    override val jbigs: List<ByteArray>
        get() {
            val ret = ArrayList<ByteArray>()

            val query = mRealm.where(JbigItem::class.java)
            val results = query.findAll()

            for (item in results) {
                if (item.jbig != null) {
                    ret.add(item.jbig!!)
                }
            }

            return ret
        }

    override fun getJbig(position: Int): ByteArray? {
        val query = mRealm.where(JbigItem::class.java)
        val results = query.findAll()

        val item = results[position]

        if (item != null) {
            return item.jbig
        }

        return null
    }

    override fun put(jbig: ByteArray) {
        Log.d("ReamlDb", "save realm object.")

        mRealm.beginTransaction()

        val item = mRealm.createObject(JbigItem::class.java)
        item.tag = "PaintView"
        item.jbig = jbig

        mRealm.commitTransaction()
    }

    override fun put(jbigs: Collection<ByteArray>) {
        for (item in jbigs) {
            put(item)
        }
    }

    override fun delete(jbig: ByteArray) {
        val query = mRealm.where(JbigItem::class.java)
        val results = query.findAll()

        for (item in results) {
            if (item.jbig == jbig) {
                mRealm.beginTransaction()
                item.deleteFromRealm()
                mRealm.commitTransaction()
                break
            }
        }
    }

    override fun delete(position: Int) {
        val query = mRealm.where(JbigItem::class.java)
        val results = query.findAll()

        results.removeAt(position)

        mRealm.commitTransaction()
    }

    override fun deleteAll() {
        val query = mRealm.where(JbigItem::class.java)
        val results = query.findAll()

        results.clear()
        mRealm.commitTransaction()
    }
}
